### ISSUE TITLE
`F821` on use of `Literal` with strings, e.g. `Literal['foo']`

### ISSUE THREAD

**Issue by samuelcolvin** [charliermarsh/ruff#128]
Since #119 was fixed via #125, I tried ruff again (build from `main`) with pydantic.

I get a few remaining `F821` errors due to use of strings in `Literal`, the line numbers are also incorrect - seems to always be `1`.

From [this](https://github.com/pydantic/pydantic/blob/f341049b9e5538a125751d75b4e44c1609b53df6/pydantic/config.py#L65-L66) code (and some other places in that file), I get:

```
pydantic/config.py:1:1: F821 Undefined name `deep`
pydantic/config.py:1:1: F821 Undefined name `none`
pydantic/config.py:1:1: F821 Undefined name `deep`
pydantic/config.py:1:1: F821 Undefined name `shallow`
pydantic/config.py:1:1: F821 Undefined name `before_validation`
pydantic/config.py:1:1: F821 Undefined name `after_validation`
pydantic/config.py:1:1: F821 Undefined name `none`
pydantic/config.py:1:1: F821 Undefined name `shallow`
pydantic/config.py:1:1: F821 Undefined name `before_validation`
pydantic/config.py:1:1: F821 Undefined name `after_validation`
```

Example code:

```py
    copy_on_model_validation: Literal['none', 'deep', 'shallow']
    post_init_call: Literal['before_validation', 'after_validation']
```

**Comment by samuelcolvin** [comment]
I have no idea how you differentiate in a general way between the following:

---
```py
from typing import Literal

def foo(x: Literal['int']):
    pass
```
---
```py
from typing import List

def foo(x: List['int']):
    pass
```
---
```py
from .local_types import CustomThing

def foo(x: CustomThing['int']):
    pass
```
---

Obviously you can whitelist `Literal`, but I have no idea how you would work out whether `CustomThing` is parameterised with strings or types which might be strings.

I guess you have to assume it's parameterised with types, can maybe one day allow a magic comment or magic property which tells ruff that `CustomThing` is parameterised with literals - maybe there's a pep for this, but I don't know of it.

**Comment by charliermarsh** [comment]
Yeah it looks like PyFlakes special-cases `Literal`: https://github.com/PyCQA/pyflakes/blob/master/pyflakes/checker.py#L1464. Will fix.

PyFlakes actually has a lot of custom logic for caching forward-reference annotation in `cast` and `TypeVar` calls which we're yet to handle, so this may pop up there too.

**Comment by vianmixtkz** [comment]
I still have the same issue
![image](https://user-images.githubusercontent.com/122097976/229777092-644cc36b-5509-4e38-bb73-5fba4bc4c5a9.png)

What version of ruff fixes it ?

I am using ruff==0.0.254

**Comment by charliermarsh** [comment]
@vianmixtkz - This was fixed long long ago. How are you importing `Literal`?

**Comment by charliermarsh** [comment]
If you add that module to [`typing-modules`](https://beta.ruff.rs/docs/settings/#typing-modules), that should fix the issue for you without having to import from `typing` directly. We don't trace imports across files, so if you import `Literal` in one file, export it, and then import that export elsewhere, we don't know that it's equivalent to `typing.Literal`. By adding your first-party module to `typing-modules`, you can tell Ruff that anything imported from that module should be considered equivalent to a `typing` import.
